File: 74-search-a-2d-matrix/search-a-2d-matrix.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
      let rowLeft = 0;
  let rowRight = matrix.length - 1;

  while (rowLeft <= rowRight) {
    let rowMid = Math.floor((rowLeft + rowRight) / 2);
    if (
      target <= matrix[rowMid][matrix[rowMid].length - 1] &&
      target >= matrix[rowMid][0]
    ) {
      let left = 0;
      let right = matrix[rowMid].length - 1;
      while (left <= right) {
        let mid = Math.floor((left + right) / 2);
        if (target === matrix[rowMid][mid]) return true;
        if (target < matrix[rowMid][mid]) {
          right = mid - 1;
        }
        if (target > matrix[rowMid][mid]) {
          left = mid + 1;
        }
      }
      break;
    } else if (target < matrix[rowMid][0]) {
      rowRight = rowMid - 1;
    } else if (target > matrix[rowMid][matrix[rowMid].length - 1]) {
      rowLeft = rowMid + 1;
    }
  }
  return false;
};

File: 74-search-a-2d-matrix/search-a-2d-matrix.java
class Solution {
 public boolean searchMatrix(int[][] matrix, int target) {
    int rowLeft = 0;
    int rowRight = matrix.length - 1;

    while (rowLeft <= rowRight) {
      int rowMid = (rowLeft + rowRight) / 2;
      if (target <= matrix[rowMid][matrix[rowMid].length - 1] &&
          target >= matrix[rowMid][0]) {
        int left = 0;
        int right = matrix[rowMid].length - 1;
        while (left <= right) {
          int mid = (left + right) / 2;
          if (target == matrix[rowMid][mid])
            return true;
          if (target < matrix[rowMid][mid]) {
            right = mid - 1;
          } else {
            left = mid + 1;
          }
        }
        return false;
      } else if (target < matrix[rowMid][0]) {
        rowRight = rowMid - 1;
      } else {
        rowLeft = rowMid + 1;
      }
    }
    return false;
  }
}